### DISCUSSION TITLE
Admin login in frontend

### DISCUSSION THREAD

**Discussion by jannismilz** [pocketbase/pocketbase#958] [Q&A] [answered]
I'm currently using NextJS 13 and I want to log in as an admin. I have environment variables with the credentials but I'd need it in a client component. Now the problem is that I can't use it there because I can't publish the credentials... And if I create an endpoint by myself and login there with the credentials in the backend.
If I'd do it that way I'm not sure how to safe the user in the cookies. I guess I could find a way for that but it's not very clean..

Thanks in advance!

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not sure that I understand the question. The SDK can be used in 3 different ways:

1. Only client-side in the browser (LocalStorage is used by default)
2. Only server-side, for example in a node server endpoint in case of SSR
3. Mixed

If you want to authenticate as admin with credentials from env variables then you need to register the client in a server-only handler. I still haven't got the time to check the changes in next.js13 so I'm not sure how this could be done in the new version, but you could find an example SSR integration for the older versions in:
- for JS SDK 0.7.x - https://github.com/pocketbase/js-sdk#ssr-integration
- for JS SDK 0.8.0-rc - https://github.com/pocketbase/js-sdk/tree/rc#ssr-integration

**Reply by jannismilz** [reply]
Okay interesting. But how do you then authorize as an admin? Like not with user input but e.g. to login the user you need an admin account. You then write the credentials just plain in the code? 🤔

**Reply by ganigeorgiev** [reply]
For JS SDK v07.x:
```js
import PocketBase from "pocketbase";

const client = new PocketBase("http://127.0.0.1:8090");

await client.admins.authViaEmail("test@example.com", "1234567890");

// every following request will be authorized until client.authStore.clear() is called...
// you can read more about the available services in https://github.com/pocketbase/js-sdk#api-services
```

For JS SDK v0.8.0-rc:
```js
import PocketBase from "pocketbase";

const client = new PocketBase("http://127.0.0.1:8090");

await client.admins.authWithPassword("test@example.com", "1234567890");

// every following request will be authorized until client.authStore.clear() is called...
// you can read more about the available services in https://github.com/pocketbase/js-sdk/tree/rc#services
```

**Reply by jannismilz** [reply]
Yes, I know what you mean **BUT** if you put it like that the credentials are public to the world in the code. (Devtools)

**Reply by ganigeorgiev** [reply]
No, you have to register the above in a **server-only** handler. The credentials could be replaced via env variables (eg. `process.env.adminEmail` and `process.env.adminPassword`).

**Reply by jannismilz** [reply]
okay with env variables of course.
But after that I want to log in a user and for that I need the provider details from the local storage. And on the server I can't select localStorage data..

**Reply by jannismilz** [reply]
Or can I do all this in my backend and it still does put the token in the auth storage?

**Reply by ganigeorgiev** [reply]
Please check the suggested ssr integration. There is an example how to use the 2 helper cookie load functions - https://github.com/pocketbase/js-sdk#ssr-integration. After the v0.8.0 release I'll add an officail guide for nextjs13 but for now you can adapt the one for the older versions.

**Reply by jannismilz** [reply]
Ahh now I get it. Thanks for the fast response!

**Comment by greendesertsnow** [comment]
Thank you for your work. But I can't connect the dots using your documentation. I'm sorry. 
Can you please put a fully functioning TODO app with the more popular choice Nextjs, instead of Sveltekit (?), online so we can check the code an all that. How do you handle the cookies? How do you integrate with auth providers? How do you protect the API and the routes?
